base/array.jl

```diff
@@ -1531,48 +1545,74 @@ end
 """
     findfirst(A)
 
-Return the linear index of the first `true` value in `A`.
+Return the index of the first `true` value in `A`.
 Return `nothing` if no such value is found.
 To search for other kinds of values, pass a predicate as the first argument.
 
+Indices are of the same type as those returned by [`keys(A)`](@ref)
+and [`pairs(A)`](@ref).
+
 # Examples
 ```jldoctest
+julia> A = [false, false, true, false]
+4-element Array{Bool,1}:
+ false
+ false
+  true
+ false
+
+julia> findfirst(A)
+3
+
+julia> findfirst(falses(3)) == nothing
+true
+
 julia> A = [false false; true false]
 2×2 Array{Bool,2}:
  false  false
   true  false
 
 julia> findfirst(A)
-2
-
-julia> findfirst(falses(3)) == nothing
-true
+CartesianIndex(2, 1)
 ```
 """
-findfirst(A) = findnext(A, 1)
+findfirst(A) = isempty(A) ? nothing : findnext(A, first(keys(A)))
```

Comment:
Calling `isempty` seems to be necessary for full generality, since there is no guaranty that `keys` returns a range, and `first` could throw an error for non-ranges. Yet it's quite annoying to have to do that. Actually, this should be done in `findnext`, but I wonder about the performance impact, given that `findnext` isn't likely to be inlined (so the check will be performed on every call).

That's related to https://github.com/JuliaLang/julia/pull/25385. Maybe we should require that `keys` returns something on which `first` never fails for numeric indices?

Following that idea, I've added a `first(::EachStringIndex) = 1` below for consistency with ranges.

Reply:
If `findfirst` is going to be generalized this much, it probably makes sense to rewrite it to iterate over `pairs` instead of calling `findnext`.

Reply:
Of course, my focus on `findnext`/`findprev` made me forget the big picture. I've added a commit to use the most generic approach, similar to `find`.